
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>TURNAP User Guide &#8212; turnap 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="turnap-user-guide">
<h1>TURNAP User Guide<a class="headerlink" href="#turnap-user-guide" title="Permalink to this headline">¶</a></h1>
<p>TURNAP (Tool for Uniform RNA Phenotyping) is a framework for generating molecular phenotypes from RNA-Seq. Existing tools have been developed to generate different transcriptomic phenotypes, and TURNAP allows you to run them with full flexibility in a convenient and organized process. The goal is to make molecular QTL mapping and other analyses as easy to perform for a multitude of transcriptomic phenotypes as for basic gene expression only.</p>
<p>A project is specified with a config file containing general parameters, e.g. paths to FASTQ and reference annotation file, single-end or paired-end, and strandedness. It also lists the phenotypes to generate.</p>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
</section>
<section id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<p>The commands and code to run the steps are given in <a class="reference external" href="https://snakemake.github.io/">Snakemake</a> files, located by default in <code class="docutils literal notranslate"><span class="pre">steps/</span></code>. This package includes sensible defaults for a set of tools, but you can edit the commands and add new phenotypes.</p>
<p>The goal of a project is to generate phenotype tables for a set of samples. Each tool could involve any combination of:</p>
<ul class="simple">
<li><p>Setup like indexing or annotation filtering</p></li>
<li><p>Processing each sample’s BAM file</p></li>
<li><p>Intermediate steps requiring data from all samples, e.g. clustering</p></li>
<li><p>Final assembly of all results into a phenotype table</p></li>
</ul>
<p>Each file in <code class="docutils literal notranslate"><span class="pre">steps/</span></code> contains all of these steps for one phenotype. Snakemake looks at the inputs and outputs for each rule and figures out the order and number of times to run each one.</p>
</section>
<section id="config-file">
<h2>Config file<a class="headerlink" href="#config-file" title="Permalink to this headline">¶</a></h2>
<p>The config file is a file in YAML format specifying general parameters, input files, and phenotypes to generate. This is the default config:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># Raw RNA-Seq data</span><span class="w"></span>
<span class="nt">fastq_map</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">data/geuvadis/fastq_map.txt</span><span class="w"></span>
<span class="nt">fastq_dir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">data/geuvadis</span><span class="w"></span>
<span class="nt">paired_end</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"></span>
<span class="nt">read_length</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">75</span><span class="w"></span>

<span class="c1"># Reference files</span><span class="w"></span>
<span class="nt">ref_genome</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">data/human_ref/Homo_sapiens.GRCh38.dna.primary_assembly.fa</span><span class="w"></span>
<span class="nt">ref_anno</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">data/human_ref/Homo_sapiens.GRCh38.106.gtf</span><span class="w"></span>
<span class="nt">retro_anno</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">data/human_ref/retro.hg38.v1.nochr.gtf</span><span class="w"></span>

<span class="c1"># Project</span><span class="w"></span>
<span class="nt">project_dir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">test</span><span class="w"></span>
<span class="nt">ref_dir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">test/reference</span><span class="w"></span>
<span class="nt">threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">4</span><span class="w"></span>

<span class="c1"># Phenotypes</span><span class="w"></span>
<span class="nt">phenotypes</span><span class="p">:</span><span class="w"></span>
<span class="nt">expression</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="nt">files</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">expression.log2.bed</span><span class="w"></span>
<span class="w">   </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">expression.tpm.bed</span><span class="w"></span>
<span class="nt">splicing</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="nt">files</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">splicing.bed</span><span class="w"></span>
<span class="nt">stability</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="nt">files</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">stability.bed</span><span class="w"></span>
<span class="nt">retroelements</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="nt">files</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">retroelements.bed</span><span class="w"></span>
<span class="nt">latent</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="nt">files</span><span class="p">:</span><span class="w"></span>
<span class="w">   </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">latent.bed</span><span class="w"></span>
</pre></div>
</div>
</section>
<section id="input-files">
<h2>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h2>
<section id="fastq-sample-map">
<h3>FASTQ sample map<a class="headerlink" href="#fastq-sample-map" title="Permalink to this headline">¶</a></h3>
<p>A file containing FASTQ file paths and the name of the sample they map to, separated by tabs, with no header.</p>
<p>Example for single-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">batch1</span><span class="o">/</span><span class="n">sampleA_run1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">sampleA</span>
<span class="n">batch1</span><span class="o">/</span><span class="n">sampleA_run2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">sampleA</span>
<span class="n">batch2</span><span class="o">/</span><span class="n">sampleA_run3</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">sampleA</span>
<span class="n">batch1</span><span class="o">/</span><span class="n">sampleB_run1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">sampleB</span>
<span class="o">...</span>
</pre></div>
</div>
<p>Example for paired-end reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">batch1</span><span class="o">/</span><span class="n">sampleA_run1_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>       <span class="n">batch1</span><span class="o">/</span><span class="n">sampleA_run1_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">sampleA</span>
<span class="n">batch1</span><span class="o">/</span><span class="n">sampleA_run2_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>       <span class="n">batch1</span><span class="o">/</span><span class="n">sampleA_run2_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">sampleA</span>
<span class="n">batch2</span><span class="o">/</span><span class="n">sampleA_run3_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>       <span class="n">batch2</span><span class="o">/</span><span class="n">sampleA_run3_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">sampleA</span>
<span class="n">batch1</span><span class="o">/</span><span class="n">sampleB_run1_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>       <span class="n">batch1</span><span class="o">/</span><span class="n">sampleB_run1_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">sampleB</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
<section id="fastq-files">
<h3>FASTQ files<a class="headerlink" href="#fastq-files" title="Permalink to this headline">¶</a></h3>
<p>FASTQ files must be compressed (decompressible with <code class="docutils literal notranslate"><span class="pre">zcat</span></code>) unless you modify the <code class="docutils literal notranslate"><span class="pre">STAR</span></code> command to handle uncompressed files. Multiple FASTQ files (or paired-end file pairs) can be provided for each sample and all reads will be used.</p>
</section>
<section id="reference-genome-sequence">
<h3>Reference genome sequence<a class="headerlink" href="#reference-genome-sequence" title="Permalink to this headline">¶</a></h3>
<p>The FASTA file for a reference genome, e.g. <code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.dna.primary_assembly.fa</span></code>.</p>
</section>
<section id="reference-annotation-file">
<h3>Reference annotation file<a class="headerlink" href="#reference-annotation-file" title="Permalink to this headline">¶</a></h3>
<p>A GTF file containing the gene, exon, and other annotations, compatible with the supplied reference genome. For example, <code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.106.gtf</span></code>.</p>
</section>
</section>
<section id="running-from-the-command-line">
<h2>Running from the command line<a class="headerlink" href="#running-from-the-command-line" title="Permalink to this headline">¶</a></h2>
<p>Phenotypes are computed by running one subprocess per phenotype category per sample. Within the specified output directory, one directory will be created for each phenotype category, containing intermediate (often per-sample) files from each program.</p>
<p>These data will then be combined into one BED file per phenotype group so that QTL mapping can be run separately for each.</p>
</section>
<section id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<section id="bed-output-files">
<h3>BED output files<a class="headerlink" href="#bed-output-files" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> files include four annotation columns, <code class="docutils literal notranslate"><span class="pre">chrom</span></code>, <code class="docutils literal notranslate"><span class="pre">chromStart</span></code>, <code class="docutils literal notranslate"><span class="pre">chromEnd</span></code>, and <code class="docutils literal notranslate"><span class="pre">name</span></code>, followed by one column per sample containing the phenotype values. There may be one phenotype per gene (e.g. expression) or multiple (e.g. splicing), but in either case the coordinates indicate the transcription start site of the phenotype’s gene. This ensures that the same cis-window variants are tested for all phenotypes of the same gene.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chrom       chromStart      chromEnd        name    HG00315 HG00106 NA18489</span>
<span class="mi">1</span>    <span class="mi">29569</span>   <span class="mi">29570</span>   <span class="n">ENSG00000227232</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">17055</span><span class="p">:</span><span class="mi">17233</span><span class="p">:</span><span class="n">clu_1_</span><span class="o">-</span>   <span class="mf">0.671533</span>        <span class="mf">0.654321</span>        <span class="mf">0.673716</span>
<span class="mi">1</span>    <span class="mi">29569</span>   <span class="mi">29570</span>   <span class="n">ENSG00000227232</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">17055</span><span class="p">:</span><span class="mi">17606</span><span class="p">:</span><span class="n">clu_1_</span><span class="o">-</span>   <span class="mf">0.0340633</span>       <span class="mf">0.037037</span>        <span class="mf">0.0241692</span>
<span class="mi">1</span>    <span class="mi">29569</span>   <span class="mi">29570</span>   <span class="n">ENSG00000227232</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">17368</span><span class="p">:</span><span class="mi">17606</span><span class="p">:</span><span class="n">clu_1_</span><span class="o">-</span>   <span class="mf">0.294404</span>        <span class="mf">0.308642</span>        <span class="mf">0.302115</span>
<span class="mi">1</span>    <span class="mi">778668</span>  <span class="mi">778669</span>  <span class="n">ENSG00000228327</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">729804</span><span class="p">:</span><span class="mi">729898</span><span class="p">:</span><span class="n">clu_2_</span><span class="o">-</span> <span class="mf">0.275362</span>        <span class="mf">0.203125</span>        <span class="mf">0.247748</span>
<span class="mi">1</span>    <span class="mi">778668</span>  <span class="mi">778669</span>  <span class="n">ENSG00000228327</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">729804</span><span class="p">:</span><span class="mi">733307</span><span class="p">:</span><span class="n">clu_2_</span><span class="o">-</span> <span class="mf">0.362319</span>        <span class="mf">0.265625</span>        <span class="mf">0.166667</span>
<span class="mi">1</span>    <span class="mi">778668</span>  <span class="mi">778669</span>  <span class="n">ENSG00000228327</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">729804</span><span class="p">:</span><span class="mi">736713</span><span class="p">:</span><span class="n">clu_2_</span><span class="o">-</span> <span class="mi">0</span>       <span class="mf">0.234375</span>        <span class="mf">0.0765766</span>
<span class="mi">1</span>    <span class="mi">778668</span>  <span class="mi">778669</span>  <span class="n">ENSG00000228327</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">733213</span><span class="p">:</span><span class="mi">733307</span><span class="p">:</span><span class="n">clu_2_</span><span class="o">-</span> <span class="mi">0</span>       <span class="mi">0</span>       <span class="mf">0.238739</span>
<span class="mi">1</span>    <span class="mi">778668</span>  <span class="mi">778669</span>  <span class="n">ENSG00000228327</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">736619</span><span class="p">:</span><span class="mi">736713</span><span class="p">:</span><span class="n">clu_2_</span><span class="o">-</span> <span class="mf">0.224638</span>        <span class="mf">0.296875</span>        <span class="mf">0.175676</span>
<span class="mi">1</span>    <span class="mi">778668</span>  <span class="mi">778669</span>  <span class="n">ENSG00000228327</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">736619</span><span class="p">:</span><span class="mi">740129</span><span class="p">:</span><span class="n">clu_2_</span><span class="o">-</span> <span class="mf">0.137681</span>        <span class="mi">0</span>       <span class="mf">0.0945946</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
</section>
<section id="running">
<h2>Running<a class="headerlink" href="#running" title="Permalink to this headline">¶</a></h2>
<p>We suggest these steps to run TURNAP.</p>
<section id="run-the-included-test-data">
<h3>1. Run the included test data<a class="headerlink" href="#run-the-included-test-data" title="Permalink to this headline">¶</a></h3>
<p>This will require installation of all programs used in the snakefiles. <a class="reference external" href="https://docs.conda.io/en/latest/miniconda.html">Miniconda</a> is recommended for easy installation and management of all these programs. Once you think you have everything installed, try running on the included test data, which is small:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>snakemake -s TURNAP/Snakefile -c1
</pre></div>
</div>
</section>
<section id="run-with-your-data">
<h3>1. Run with your data<a class="headerlink" href="#run-with-your-data" title="Permalink to this headline">¶</a></h3>
<p>Snakemake has features to handle many execution needs such as threads, computational resources, and automatic cluster job submission.</p>
<div class="toctree-wrapper compound">
</div>
</section>
</section>
<section id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">turnap</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Daniel Munro.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.5.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>